integration/kotlinx-coroutines-play-services/src/Tasks.kt

```diff
@@ -44,45 +37,97 @@ public fun <T> Deferred<T>.asTask(): Task<T> {
 
 /**
  * Converts this task to an instance of [Deferred].
+ *
+ * Prefer passing the corresponding [CancellationTokenSource] if the [Task] can be created with a [CancellationToken]
+ * to support bi-directional cancellation.
+ *
  * If task is cancelled then resulting deferred will be cancelled as well.
  */
-public fun <T> Task<T>.asDeferred(): Deferred<T> {
+public fun <T> Task<T>.asDeferred(): Deferred<T> = asDeferred(CancellationTokenSource())
+
+/**
+ * Converts this task to an instance of [Deferred] with a [CancellationTokenSource] to control cancellation.
+ *
+ * If the task is cancelled, then the [cancellationTokenSource] and the resulting deferred will be cancelled.
+ * If the deferred is cancelled, then the [cancellationTokenSource] will be cancelled.
+ * If the [cancellationTokenSource] is cancelled, then the deferred will be cancelled.
+ */
+@ExperimentalCoroutinesApi
+public fun <T> Task<T>.asDeferred(cancellationTokenSource: CancellationTokenSource): Deferred<T> {
+    val deferred = CompletableDeferred<T>()
+
     if (isComplete) {
         val e = exception
-        return if (e == null) {
-            @Suppress("UNCHECKED_CAST")
-            CompletableDeferred<T>().apply { if (isCanceled) cancel() else complete(result as T) }
+        if (e == null) {
+            if (isCanceled) {
+                deferred.cancel()
+            } else {
+                @Suppress("UNCHECKED_CAST")
+                deferred.complete(this.result as T)
+            }
         } else {
-            CompletableDeferred<T>().apply { completeExceptionally(e) }
+            deferred.completeExceptionally(e)
+        }
+    } else if (cancellationTokenSource.token.isCancellationRequested) {
+        // The task hasn't completed, yet cancellation was already requested.
+        // Interpret this by cancelling immediately (no way to cancel the task)
+        deferred.cancel()
+    } else {
+        addOnCompleteListener {
+            val e = it.exception
+            if (e == null) {
+                @Suppress("UNCHECKED_CAST")
+                if (it.isCanceled) deferred.cancel() else deferred.complete(it.result as T)
+            } else {
+                deferred.completeExceptionally(e)
+            }
+        }
+
+        cancellationTokenSource.token.onCanceledRequested {
+            deferred.cancel()
         }
     }
 
-    val result = CompletableDeferred<T>()
-    addOnCompleteListener {
-        val e = it.exception
-        if (e == null) {
-            @Suppress("UNCHECKED_CAST")
-            if (isCanceled) result.cancel() else result.complete(it.result as T)
-        } else {
-            result.completeExceptionally(e)
+    deferred.invokeOnCompletion {
+        if (it is CancellationException) {
+            cancellationTokenSource.cancel()
         }
     }
-    return result
+
+    return deferred
```

Comment:
```suggestion
    return object : Deferred<T> by deferred {}
```

so no casting will occur